lib/semantics/resolve-names-utils.cc

```diff
@@ -286,4 +286,158 @@ Bound ArraySpecAnalyzer::GetBound(const parser::SpecificationExpr &x) {
+  const auto &name{std::get<parser::Name>(stmt.statement.t)};
+  return BuildSubprogramTree(name, x).set_stmt(stmt).set_endStmt(end);
+}
+ProgramTree ProgramTree::Build(const parser::SubroutineSubprogram &x) {
+  const auto &stmt{std::get<parser::Statement<parser::SubroutineStmt>>(x.t)};
+  const auto &end{std::get<parser::Statement<parser::EndSubroutineStmt>>(x.t)};
+  const auto &name{std::get<parser::Name>(stmt.statement.t)};
+  return BuildSubprogramTree(name, x).set_stmt(stmt).set_endStmt(end);
+}
+ProgramTree ProgramTree::Build(const parser::SeparateModuleSubprogram &x) {
+  const auto &stmt{std::get<parser::Statement<parser::MpSubprogramStmt>>(x.t)};
+  const auto &end{
+      std::get<parser::Statement<parser::EndMpSubprogramStmt>>(x.t)};
+  const auto &name{stmt.statement.v};
+  return BuildSubprogramTree(name, x).set_stmt(stmt).set_endStmt(end);
+}
+ProgramTree ProgramTree::Build(const parser::Module &x) {
+  const auto &stmt{std::get<parser::Statement<parser::ModuleStmt>>(x.t)};
+  const auto &end{std::get<parser::Statement<parser::EndModuleStmt>>(x.t)};
+  const auto &name{stmt.statement.v};
+  return BuildModuleTree(name, x).set_stmt(stmt).set_endStmt(end);
+}
+ProgramTree ProgramTree::Build(const parser::Submodule &x) {
+  const auto &stmt{std::get<parser::Statement<parser::SubmoduleStmt>>(x.t)};
+  const auto &end{std::get<parser::Statement<parser::EndSubmoduleStmt>>(x.t)};
+  const auto &name{std::get<parser::Name>(stmt.statement.t)};
+  return BuildModuleTree(name, x).set_stmt(stmt).set_endStmt(end);
+}
+ProgramTree ProgramTree::Build(const parser::BlockData &x) {
+  DIE("BlockData not yet implemented");
+}
+
+const parser::ParentIdentifier &ProgramTree::GetParentId() const {
+  const auto *stmt{
+      std::get<const parser::Statement<parser::SubmoduleStmt> *>(stmt_)};
+  return std::get<parser::ParentIdentifier>(stmt->statement.t);
+}
+
+bool ProgramTree::IsModule() const {
+  auto kind{GetKind()};
+  return kind == Kind::Module || kind == Kind::Submodule;
+}
+
+Symbol::Flag ProgramTree::GetSubpFlag() const {
+  return GetKind() == Kind::Function ? Symbol::Flag::Function
+                                     : Symbol::Flag::Subroutine;
+}
+
+bool ProgramTree::HasModulePrefix() const {
+  using ListType = std::list<parser::PrefixSpec>;
+  const auto *prefixes{std::visit(
+      common::visitors{
+          [](const parser::Statement<parser::FunctionStmt> *x) {
+            return &std::get<ListType>(x->statement.t);
+          },
+          [](const parser::Statement<parser::SubroutineStmt> *x) {
+            return &std::get<ListType>(x->statement.t);
+          },
+          [](const auto *) -> const ListType * { return nullptr; },
+      },
+      stmt_)};
+  if (prefixes) {
+    for (const auto &prefix : *prefixes) {
+      if (std::holds_alternative<parser::PrefixSpec::Module>(prefix.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+ProgramTree::Kind ProgramTree::GetKind() const {
```

Comment:
This could be implemented as a `constexpr` templatized variable, too.

```
template<typename A> constexpr ProgramTree::Kind /*or whatever the type is */ KindOfStmt;
template<> constexpr ProgramTree::Kind<parser::Statement<parser::ProgramStmt>> KindOfStmt{ProgramTree::Kind::Program};
// &c.
return std::visit([](const auto *x){ return KindOfStmt<std::decay_t<decltype(*x)>>; }, stmt_);
```
which may or may not be as clear but would be fewer lines of code.

Reply:
My attempt at using this approach is below. I don't think it's clearer or shorter.
```
template<typename A> constexpr ProgramTree::Kind KindOfStmt;
template<>
constexpr ProgramTree::Kind KindOfStmt<parser::Statement<parser::ProgramStmt>>{
    ProgramTree::Kind::Program};
template<>
constexpr ProgramTree::Kind KindOfStmt<parser::Statement<parser::FunctionStmt>>{
    ProgramTree::Kind::Function};
template<>
constexpr ProgramTree::Kind
    KindOfStmt<parser::Statement<parser::SubroutineStmt>>{
        ProgramTree::Kind::Subroutine};
template<>
constexpr ProgramTree::Kind
    KindOfStmt<parser::Statement<parser::MpSubprogramStmt>>{
        ProgramTree::Kind::MpSubprogram};
template<>
constexpr ProgramTree::Kind KindOfStmt<parser::Statement<parser::ModuleStmt>>{
    ProgramTree::Kind::Module};
template<>
constexpr ProgramTree::Kind
    KindOfStmt<parser::Statement<parser::SubmoduleStmt>>{
        ProgramTree::Kind::Submodule};

ProgramTree::Kind ProgramTree::GetKind() const {
  return std::visit(
      [](const auto *x) { return KindOfStmt<std::decay_t<decltype(*x)>>; },
      stmt_);
```